api/utils/uaaClient.js

```diff
@@ -1,43 +1,300 @@
 const request = require('request');
-const url = require('url');
 const config = require('../../config');
 
+const uaaOptions = config.passport.uaa;
+
+/**
+ * @typedef {object} UAAInvite
+ * @property {string} email
+ * @property {string} userId
+ * @property {string} origin
+ * @property {boolean} success
+ * @property {string=} errorCode
+ * @property {string=} errorMessage
+ * @property {string=} inviteLink
+ */
+
 class UAAClient {
-  constructor(accessToken) {
-    this.accessToken = accessToken;
+  constructor() {
+    this.clientId = uaaOptions.options.clientID;
+    this.clientSecret = uaaOptions.options.clientSecret;
   }
 
-  async verifyUserGroup(userId, groupNames = []) {
-    const { groups, origin, verified } = await this.request('GET', `/Users/${userId}`);
+  /**
+   *
+   * Public API
+   *
+   * These methods encapsulate some business logic and make multiple requests
+   *
+   */
+
+  /**
+   * @param {string} targetUserEmail - the email address of the user to add
+   * @param {string} userToken - a user token with the `scim.invite` scope
+   *
+   * Invites the target user to UAA and adds them to the `pages.user` UAA group
+   */
+  async inviteUserToUserGroup(targetUserEmail, userToken) {
+    const groupName = 'pages.user';
+
+    const clientToken = await this.fetchClientToken();
+
+    const uaaUser = await this.fetchUserByEmail(targetUserEmail, clientToken);
+    const isInGroup = uaaUser && this.userInGroup(uaaUser.groups, [groupName]);
+    if (isInGroup) {
+      return null;
+    }
+
+    const invite = uaaUser ? null : await this.inviteUser(targetUserEmail, userToken);
+
+    const { origin } = invite || uaaUser;
+    const userId = (invite && invite.userId) || uaaUser.id;
+
+    const groupId = await this.fetchGroupId(groupName, clientToken);
+    await this.addUserToGroup(groupId, { origin, userId }, userToken);
+
+    return invite;
+  }
+
+  /**
+   * @param {string} userId - a UAA user id
+   * @param {[string]} groupNames - allowed UAA group names, ex: ['pages.user']
+   *
+   * Verifies that the UAA user is in the specified UAA group
+   */
+  async verifyUserGroup(userId, groupNames) {
+    const clientToken = await this.fetchClientToken();
+    const { groups, origin, verified } = await this.fetchUser(userId, clientToken);
+
     if (origin === 'cloud.gov' && !verified) {
       return false;
     }
 
-    return groups.filter(group => groupNames.includes(group.display)).length > 0;
+    return this.userInGroup(groups, groupNames);
+  }
+
+  /**
+   *
+   * Utility
+   *
+   */
+
+  /**
+   * @param {[{display: string}]} userGroups - groups that the user belongs to
+   * @param {[string]} groupNames - allowed UAA group names, ex: ['pages.user']
+   */
+  userInGroup(userGroups, groupNames) {
```

Comment:
method name userInGroups (plural) would be helpful ... i had to do a double take ... 
or could have both userInGroup and userInGroups methods?

Reply:
ha, yeah, I dunno. We're checking to see if the user is in at least one group of the list provided. I could def go either way.

Reply:
yeah ... userInAtLeastOneGroup is a bit wordy 😄 
i think right now it requires a dev to already know this function to use it